[llvm-lit] Print environment variables when using env without subcommand

Harini0924 opened the pull request:
This patch addresses an issue with lit's internal shell when env is without any arguments, it fails with exit code 127 because `env` requires a subcommand. This patch addresses the issue by encoding the command to properly return environment variables even when no arguments are provided. 

The error occurred when running the command 
` LIT_USE_INTERNAL_SHELL=1 ninja check-llvm`.

fixes: #102383
This is part of the test cleanups proposed in the RFC: [[RFC] Enabling the Lit Internal Shell by Default](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179)

Harini0924 committed:
[llvm-lit]

When executing not --crash env without any arguments,
it fails with exit code 127 because env reuires a
subcommand. This path addresses the issue by encoding
the command to properly return environment variables
even when no arguments are provided.

Harini0924 commented:
CC: @ilovepi @petrhosek

ilovepi reviewed (approved):
LGTM, but let’s modify the commit message slightly. First, it’s not really important to include the context about the “not —crash” bits. Additionally, the precise error code isn’t important only that the command failed. 

This patch is really about the internal shell’s support for printing the environment as output. So let’s trim it down to just those bits.

petrhosek commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -743,7 +743,8 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
                 if not args:
-                    raise InternalShellError(j, "Error: 'env' requires a" " subcommand")
+                    # Return the environment variables if no argument is provided 
+                    return {key: value for key, value in cmd_shenv.env.items()}
```
I don't think this is correct since there might be other piped commands, for example you could do this:
```
env | grep TERM
```
Furthermore all return paths from this function return exit code so returning a dictionary here is not correct.

jh7370 reviewed (commented):
What's the purpose of doing `not --crash env`? That sounds to me like you're trying to show that the internal lit `env` doesn't crash?

Assuming it is needed, shouldn't there be a test for this?

ilovepi commented:
> What's the purpose of doing `not --crash env`? That sounds to me like you're trying to show that the internal lit `env` doesn't crash?
> 
That comes from https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/not/disable-symbolization.test, so I'd guess its more about `not` than `env`. I'd hazard that test could be rewritten w/o `env`, but I think some other tests outside of llvm also rely on this behavior.

> Assuming it is needed, shouldn't there be a test for this?

Actually yes, I agree. I confused this case w/ another one that was NFC. 

@Harini0924, let's add a test for the new `env` behavior. Those should be updated in https://github.com/llvm/llvm-project/blob/main/llvm/utils/lit/tests/shtest-env.py. The fact that that test still seems to be passing means that we're not handling this case correctly. Likely it still fails, but does that differently.

ilovepi reviewed (changes requested):
I should have checked a bit more closely when reviewing this the first time. I assumed when the LLVM tests had passed, that this change was doing the right thing, but as pointed out by other reviewers, it is not.

So, please add some testing for `env`, and instead of returning the dictionary, we can  just call a function to correctly print the environment instead.

Harini0924 committed:
[llvm-lit] Resolve env subcommand required error

This patch addresses an issue where calling env without any arguments incorrectly returns
a "requires subcommand" message. The patch modifies the env command to properly display
the current environment variables even when no arguments are provided. Additionally,
the test cases have been updated: instead of returning an error, the tests now pass
and check if the environment variables are output correctly.

Harini0924 committed:
[llvm-lit] Resolve env subcommand required error

This patch addresses an issue where calling env without any arguments incorrectly returns
a "requires subcommand" message. The patch modifies the env command to properly display
the current environment variables even when no arguments are provided. Additionally,
the test cases have been updated: instead of returning an error, the tests now pass
and check if the environment variables are output correctly.

Harini0924 committed:
[llvm-lit] Resolve env subcommand required error

This patch addresses an issue where calling env without any arguments incorrectly returns
a "requires subcommand" message. The patch modifies the env command to properly display
the current environment variables even when no arguments are provided. Additionally, the
test cases have been updated: instead of returning an error, the tests now pass and check
if the environment variables are output correctly.

Harini0924 committed:
hi

Harini0924 committed:
[llvm-lit] Resolve env subcommand required error

This patch addresses an issue where calling env without any arguments incorrectly returns a "requires subcommand" message. The patch modifies the env command to properly display the current environment variables even when no arguments are provided. Additionally, the test cases have been updated: instead of returning an error, the tests now pass and check if the environment variables are output correctly.

Harini0924 committed:
[llvm-lit] Resolve env subcommand required error

This patch addresses an issue where calling env without any arguments incorrectly
returns a "requires subcommand" message. The patch modifies the env command to
properly display the current environment variables even when no arguments are provided.
Additionally, the test cases have been updated: instead of returning an error, the tests
now pass and check if the environment variables are output correctly.

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-env/env-args-none.txt:
```diff
@@ -1 +1,30 @@
-# RUN: env
+# Check Default Environment
+# RUN: env | FileCheck -check-prefix=CHECK-2-EMPTY %s
```
Lets update the prefix here. In the example you drew from there were 2 env uses, so we don't need the `2`. I'd also say you don't need the `CHECK- ` prefix. These comments should apply to all the prefixes in the file.

So, something like the following seems a bit better.
```suggestion
# RUN: env | FileCheck -check-prefix=EMPTY %s
```

jh7370 replied:
Please don't use any check prefix ending in `EMPTY`, because FileCheck has an `EMPTY` suffix for checking if a line is completely empty.

ilovepi commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -744,7 +744,9 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
-                    return {key: value for key, value in cmd_shenv.env.items()}
+                    env_str = "\n".join(f"{key}={value}" for key, value in cmd_shenv.env.items())
+                    results.append(ShellCommandResult(j, env_str, "", 0 , timeoutHelper.timeoutReached(), []))
+                    return 0
```
I'm still not 100% that we should be returning here. Have you tested this on the rest of the project? So not just the unit tests, but by running something like `check-all` w/ the internal shell enabled?

Harini0924 replied:
@ilovepi @petrhosek I made changes to the handling of the `env` command when no arguments are provided to ensure compatibility with piped commands and maintain the correct exit codes. Previously, the code raised an error if env had no arguments, which would break commands like `env | grep TERM `. By modifying the code to generate a string of environment variables and append it to the results list, the env command can now function correctly in pipelines, allowing subsequent commands to receive the expected input. Additionally, it  returns 0, signifying successful execution and maintaining consistency across different execution paths. I tested these changes by running `LIT_USE_INTERNAL_SHELL=1 ninja check-all`, and the tests pass, confirming that the modifications work across the project.

ilovepi reviewed (commented):
Thanks for adding the tests. I think this is pretty close for me, but I'm still a bit unsure about the early return. @petrhosek brought up a technical point about that, and I think we should determine if those are still a problem before landing this as is. If you can, provide some context on what you found, that may help us make that determination faster.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -743,7 +743,10 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
                 args = updateEnv(cmd_shenv, args)
                 if not args:
-                    raise InternalShellError(j, "Error: 'env' requires a" " subcommand")
+                    # Return the environment variables if no argument is provided 
```
```suggestion
                    # Return the environment variables if no argument is provided.
```
Trailing whitespace.

Also, please make sure all your python files are formatted with darker.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env/env-args-none.txt:
```diff
@@ -1 +1,30 @@
-# RUN: env
+# Check Default Environment
```
```suggestion
# Check Default Environment.
```
Nit: comments should end with a '.'. Also, does this really need a capital 'd' and 'e'? Same throughout.

Harini0924 replied:
I’ve already addressed that and made the necessary corrections. Thanks for the feedback!

ilovepi commented:
@jh7370 I think we have some other tests that have an `EMPTY` suffix, such as https://github.com/llvm/llvm-project/blob/9ba524427321b931bad156860755adf420aeec6a/llvm/utils/lit/tests/Inputs/shtest-env/env-calls-env.txt#L6

I suppose we should update those too(separately).

ilovepi reviewed (approved):
LGTM as long as bots are happy. I'll land this for you once those pass and @jh7370 is satisfied.

jh7370 reviewed (changes requested):
This still hasn't been run through the python code formatter (and there's still trailing whitespace on the line I highlighted before).

See also https://llvm.org/docs/CodingStandards.html#python-version-and-source-code-formatting.

jh7370 reviewed (commented):
The coding standard I linked last time say only to modify the bits of code you touch in the PR, not the whole files. It looks like some of these files aren't even touched by this PR...

(Sorry for being fussy)

jh7370 commented:
I think you've stuffed something up on the branch?

RoboTux commented:
There's still lots of lint changes on lines not modified by this PR. Try clang-format-diff or git-clang-format to only format lines you have changed.

ilovepi commented on llvm/utils/lit/tests/Inputs/per-test-coverage/per-test-coverage.py:
```diff
@@ -1,12 +1,11 @@
-# Check that the environment variable is set correctly
-# RUN: %{python} %s | FileCheck -DINDEX=1 %s
-# RUN: %{python} %s | FileCheck -DINDEX=2 %s
+#Check that the environment variable is set correctly
```
This file shouldn’t be modified at all. Please make it match what is on main.

Harini0924 replied:
yes I will change that right now

ilovepi commented on llvm/utils/lit/tests/Inputs/per-test-coverage-by-lit-cfg/per-test-coverage-by-lit-cfg.py:
```diff
@@ -1,12 +1,11 @@
-# Check that the environment variable is set correctly
-# RUN: %{python} %s | FileCheck -DINDEX=1 %s
-# RUN: %{python} %s | FileCheck -DINDEX=2 %s
+#Check that the environment variable is set correctly
```
This file also isn’t part of the PR. So please remove it from the diff.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -417,7 +417,7 @@ def maybeUnescape(arg):
         stdout.write(encode("\n"))
 
-    for (name, mode, f, path) in opened_files:
+    for name, mode, f, path in opened_files:
```
As stated already, you shouldn't be formatting parts of files that you aren't already changing. You haven't modified this for loop, for example, so this change shouldn't exist.

Harini0924 replied:
Thank you for your feedback. I understand the importance of modifying only the necessary parts of the files. To ensure this, I will use git diff to review my changes before committing. This will help me confirm that only the intended modifications are included and avoid altering any untouched sections.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -583,7 +583,7 @@ def processRedirects(cmd, stdin_source, cmd_shenv, opened_files):
     # where file-object is initially None.
     redirects = [(0,), (1,), (2,)]
-    for (op, filename) in cmd.redirects:
+    for op, filename in cmd.redirects:
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -605,7 +605,7 @@ def processRedirects(cmd, stdin_source, cmd_shenv, opened_files):
     # Open file descriptors in a second pass.
     std_fds = [None, None, None]
-    for (index, r) in enumerate(redirects):
+    for index, r in enumerate(redirects):
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -911,7 +920,7 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
     # on Win32, for example). Since we have already spawned the subprocess, our
     # handles have already been transferred so we do not need them anymore.
-    for (name, mode, f, path) in opened_files:
+    for name, mode, f, path in opened_files:
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -965,7 +974,7 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
         output_files = []
         if res != 0:
-            for (name, mode, f, path) in sorted(opened_files):
+            for name, mode, f, path in sorted(opened_files):
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -1130,7 +1139,7 @@ def executeScriptInternal(
 
         # Add the command output, if redirected.
-        for (name, path, data) in result.outputFiles:
+        for name, path, data in result.outputFiles:
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -1369,8 +1378,11 @@ def regex_escape(s):
     path_substitutions = [
-        ("s", sourcepath), ("S", sourcedir), ("p", sourcedir),
-        ("t", tmpName), ("T", tmpDir)
+        ("s", sourcepath),
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -1388,15 +1400,17 @@ def regex_escape(s):
         substitutions.append(("%{" + letter + ":real}", real_path))
-        substitutions.append(("%{/" + letter + ":real}",
-            real_path.replace("\\", "/")))
+        substitutions.append(("%{/" + letter + ":real}", real_path.replace("\\", "/")))
```
Same comment.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -1388,15 +1400,17 @@ def regex_escape(s):
         substitutions.append(
-            ("%{/" + letter + ":regex_replacement}",
-            regex_escape(path.replace("\\", "/"))))
+            (
```
Same comment.

jh7370 reviewed (commented):
I've been explicit about which changes to remove this time. As already explained here and in the documentation, you shouldn't modify parts of files you aren't already modifying. In theory, the tooling should support this (follow the instructions in the coding standards guide for suggestions on how to do this).

arichardson commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -743,7 +743,16 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
                 if not args:
-                    raise InternalShellError(j, "Error: 'env' requires a" " subcommand")
+                    # Return the environment variables if no argument is provided.
+                    env_str = "\n".join(
```
Is joining with LF correct for windows output? Or is this fine since we are emulating a UNIX shell for Windows?

jh7370 replied:
FWIW, I'd expect \n to be fine on its own for Windows too. FileCheck will typically canonicalise the whitespace, so I wouldn't expect it to have any impact on the output in the typical case (there might be some weird edge cases, I guess, but I'd be surprised if they impacted anything in practice).

arichardson commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -743,7 +743,16 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
-                    raise InternalShellError(j, "Error: 'env' requires a" " subcommand")
+                    # Return the environment variables if no argument is provided.
+                    env_str = "\n".join(
+                        f"{key}={value}" for key, value in cmd_shenv.env.items()
```
I believe this is not required for POSIX compatibility, but sorting these items alphabetically by key would make tests slightly more reproducible.

Harini0924 replied:
I will list the items alphabetically by key.

arichardson commented on llvm/utils/lit/tests/Inputs/shtest-env/env-args-none.txt:
```diff
@@ -1 +1,30 @@
+# Check default environment.
+# RUN: env | FileCheck -check-prefix=NO-ARGS %s
+#
+# NO-ARGS: BAR = 2
```
Did you actually run these tests, I can't see how this is passing?
It isn't using the `print_environment.py` script so there should be no spaces between the output. Additionally, I think we need to sort the variables as noted above to make this tests non-flaky.

Harini0924 replied:
I will examine this more closely right now. Thank you for your feedback

arichardson commented on llvm/utils/lit/tests/shtest-env.py:
```diff
@@ -9,35 +9,35 @@
-# CHECK: # | Error: 'env' requires a subcommand
-# CHECK: # error: command failed with exit status: {{.*}}
+# CHECK-NOT: {{^[^#]}}
+# CHECK: --
```
Is this on the next line, if so we can simplify these passing tests using `CHECK-NEXT: --` instead of the CHECK-NOT: followed by CHECK.

Harini0924 replied:
Yes, if the expected lines are sequential and should appear in the exact order without any unexpected lines in between, we can simplify these passing tests using CHECK-NEXT: instead of CHECK-NOT: followed by CHECK. I will change it right now.

jh7370 replied:
What exactly is the output that you get in these cases?

jh7370 replied:
(As an aside, I see this pattern is followed below, so I'm not suggesting you change it now).

arichardson reviewed (approved):
Two minor questions but once resolved this looks good to me.

Harini0924 committed:
[llvm-lit] Using a commit ID that passes all test cases

This will pass the test cases

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env/env-args-none.txt:
```diff
@@ -1 +1,30 @@
+# Unset environment variables.
+# RUN: env -u FOO -u BAR | FileCheck -check-prefix=UNSET-U %s
+#
+# UNSET_U-NOT: BAR
```
```suggestion
# UNSET-U-NOT: BAR
```

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env/env-args-none.txt:
```diff
@@ -1 +1,30 @@
-# RUN: env
+# Check default environment.
```
I think it would be useful to show what happens to variables that aren't set or unset on the command-line, in cases where other variables are. The easiest way to do this would be to add an additional environment variable to the default set and then just check it has its original value in every test case.

jh7370 commented:
Please review the pre-merge check failure. If you take a look, you'll see that there are STILL problems with your formatting (for example, there are two cases of trailing whitespace). Sorry to be blunt, but this shows that you have not run `darker` properly on your code, despite me asking you to several times. If you think you have, perhaps it is best to say so and then advise what you have tried.

Also, I just wanted to confirm that you've definitely run the new test and that it passes?

Harini0924 committed:
[llvm-lit] Formatting Python files

In my last commit, I attempted to format the Python files using `darker`,
but the changes were not applied correctly. This commit ensures that
`darker` has been properly used to format the code.

Harini0924 commented:
@jh7370 I apologize for the formatting issues and any inconvenience caused. Running darker has been challenging for me as I need to manually run it on the Python files, and I suspect there may have been issues with the setup. However, I believe I have correctly implemented and run darker on my latest commit, which should address the formatting concerns, including the trailing white space.

Regarding the new test, I have added a test case to show how variables that aren't set or unset on the command-line are handled. The approach was to include an additional environment variable in the default set and check that it retains its original value in every test case. This setup ensures that the additional environment variable (BAZ) is correctly handled while verifying the values of BAR and FOO. 

If there are any further issues or additional steps needed, please let me know. Thank you for your understanding and patience.

Harini0924 committed:
Resolved conflicts and added changes to env-args-none.txt, lit.cfg, and shtest-env.py

Harini0924 committed:
[llvm-lit] Add pattern to CHECK lines for env command

Added '| {{.*}}' to ensure proper matching for env command output.

Harini0924 committed:
Merge branch 'llvm:main' into llvm_env

Harini0924 committed:
[llvm-lit] Used darker to format the python files

Used darker to format the llvm/utils/lit/tests/shtest-env.py file.

Harini0924 commented:
@jh7370 Thank you so much for being patient with me. I've made some changes and would really appreciate it if you could take a look at them. If there's anything else I need to add or change, please let me know.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env/env-args-none.txt:
```diff
@@ -1,34 +1,35 @@
-# Check default environment.
+#Check default environment.
```
Please revert back to `# Check default environment.` like it was before.

jh7370 commented on llvm/utils/lit/tests/shtest-env.py:
```diff
@@ -34,9 +34,9 @@
+# CHECK: env | {{.*}}
 # CHECK: # executed command: env
-# CHECK-NOT: {{^[^#]}}
+# CHECK-NOT: {{^[^#]}} | {{.*}}
```
Could you explain the additions you've made here, please?

Harini0924 replied:
I added `# CHECK: env | {{.*}}` to make sure that the test captures the full line where the env command is executed and its output is piped to FileCheck. The `{{.*}}`matches everything after `env |`, ensuring the entire command and its arguments are included in the test output. This verifies that the env command is executed properly within the test and its output is processed as expected. 
Other tests also use `| {{.*}}` to match the full lines of commands. For example, in tests like `shtest-env :: env-u.txt`, the commands include `[[PYTHON]] print_environment.py | {{.*}}` and `env -u FOO [[PYTHON]]` `print_environment.py | {{.*}}`. This approach ensures that the entire command line and its output are correctly captured and checked in the test output.

ilovepi replied:
Well, I’m also a bit concerned about the CHECK-NOT case. It’s often hard to write a negative match that works the way you intended.  The pattern used elsewhere in this file does not have a trailing |{{.*}}.  Instead it would fail if it found any line with a leading “#”.  The new pattern does not preserve that property, and is thus would miss some cases that would otherwise be caught. 

The CHECK-NOT checks are not affected by —match-whole-lines, so there isn’t a need for a trailing {{.*}} in any case. https://llvm.org/docs/CommandGuide/FileCheck.html#cmdoption-FileCheck-match-full-lines

My suggestion would be to use that common pattern for the negative match.

Harini0924 replied:
I understand your concern about the CHECK-NOT case. To keep things consistent, I've updated the checks to cover all the executed commands for env-args-none.txt. This way, we maintain uniformity across all tests and ensure we don't miss any potential issues. By checking all the commands, we can be more confident that the tests are more accurate. Additionally, we can still use `# CHECK-NOT: {{^[^#]}}`.

Harini0924 committed:
[llvm-lit] Added a space in env-args-none.txt

Reverted back to # Check default environment.

Harini0924 committed:
[llvm-lit] Updated shtest-env.py for Consistent Output Checking

Reformatted test checks in env-args-none.txt to stay consistent with
existing passing tests.

arichardson commented on llvm/utils/lit/tests/shtest-env.py:
```diff
@@ -9,35 +9,43 @@
 # CHECK: # executed command: env FOO=1
-# CHECK: # | Error: 'env' requires a subcommand
-# CHECK: # error: command failed with exit status: {{.*}}
+# CHECK-NOT: {{^[^#]}}
```
```suggestion
# CHECK-NOT: # error:
```
Would this be easier to understand?

Harini0924 replied:
I tested out the suggested change, but encountered an error where the test could not find the expected string "# error:" in the input. Given this issue, I will continue using the current approach with # CHECK-NOT: `{{^[^#]}}`.

jh7370 replied:
I don't think what you've said makes sense. @arichardson asked for a `CHECK-NOT: # error:` not an additional `CHECK: # error:`. My concern is that you don't really understand what he suggested, which in turn suggests you don't really understand how the testing works. Is that the case?

Harini0924 replied:
I apologize for the misunderstanding. Thank you for clarifying @arichardson's request.

To address your concerns accurately, I understand that the suggestion was to use CHECK-NOT: # error: rather than an additional CHECK: # error:. This ensures that the expected output does not contain any unexpected error messages, improving the reliability and precision of the tests. I have added this to my recent commit.

Harini0924 commented:
@jh7370 I've fixed the missing space in `env-args-none.txt`. Is there anything else I should adjust or add? Please let me know.

Harini0924 committed:
[llvm-lit] Modified `CHECK-NOT: {{^[^#]}}`

Updated test cases to use `CHECK-NOT: # error`.

Harini0924 commented:
@jh7370 @arichardson I've changed `# CHECK-NOT: {{^[^#]}}` to  `# CHECK-NOT: # error:` for the tests I modified. To maintain consistency across the whole file, should update the other passing tests to use `# CHECK-NOT: # error:` as well.

jh7370 commented:
> @jh7370 @arichardson I've changed `# CHECK-NOT: {{^[^#]}}` to `# CHECK-NOT: # error:` for the tests I modified. To maintain consistency across the whole file, should update the other passing tests to use `# CHECK-NOT: # error:` as well?

Generally, I'd avoid making consistency changes in the same PR, and make them in a follow-up instead.

That being said, I'm wondering if the intent of the other lines wasn't to check specifically for an error, but rather to check for any executed command output at all? Unfortunately, I'm not really familiar with this test or its background. I'd be reluctant to change what the test is really checking for without understanding its purpose.

Harini0924 commented:
@jh7370 To clarify the changes I made: Originally, when the tests were failing, there was a check for errors using the line `# CHECK: # error: command failed with exit status: {{.*}}`. However, since the tests have been modified and are now passing, I replaced that check with `# CHECK-NOT: {{^[^#]}}`. This change reflects the fact that the tests are no longer expecting an error, but rather verifying that no unexpected output is generated.

Given this, I think it makes sense to maintain the original format for passing tests, which is `# CHECK-NOT: {{^[^#]}}`, instead of `# CHECK-NOT: # error:`. Does this approach make sense to you, or do you have any thoughts on changing this? If this makes sense, I can revert the last commit and use `# CHECK-NOT: {{^[^#]}}` consistently for the tests that pass. Please let me know if I should clarify anything further.

jh7370 commented:
> @jh7370 To clarify the changes I made: Originally, when the tests were failing, there was a check for errors using the line `# CHECK: # error: command failed with exit status: {{.*}}`. However, since the tests have been modified and are now passing, I replaced that check with `# CHECK-NOT: {{^[^#]}}`. This change reflects the fact that the tests are no longer expecting an error, but rather verifying that no unexpected output is generated.
> 
> Given this, I think it makes sense to maintain the original format for passing tests, which is `# CHECK-NOT: {{^[^#]}}`, instead of `# CHECK-NOT: # error:`. Does this approach make sense to you, or do you have any thoughts on changing this? If this makes sense, I can revert the last commit and use `# CHECK-NOT: {{^[^#]}}` consistently for the tests that pass. Please let me know if I should clarify anything further.

Just spelling out what the regex means, because I've tripped myself up at least twice trying to interpret it. `^` means "start of line". `[^#]` means "any character except `#`. Combined they mean "match a non-empty line that does not start with `#`". `CHECK-NOT` instructs FileCheck to fail if it sees the specified between the previous thing it matched and the next thing it matches, so the combined effect is to fail the test if, between the previous and next patterns, a non-empty line _without_ a leading `#` appears. In other words, it will NOT fail if it sees e.g. `# error:` in the output, because of the leading `#` (and based on the checks elsewhere in the file, I assume that is indeed the format that errors will be printed in). Could somebody confirm that I'm not making a mistake?

@Harini0924, if you didn't have any other tests to base things on, and you wanted to write this specific test completely from scratch, what would your CHECK lines look like? The way you've been replying, I feel like you don't really have a clear idea of what you are trying to achieve, which makes me nervous. On the other hand, maybe it's just a communication thing.

---

Unrelated, but "Resolve env subcommand required error" isn't really describing what this patch does in a useful manner. I think the key thing is that you're changing env without subcommands to print the environment variables, so I'd change this to "
[llvm-lit] Print environment variables when using env without subcommand".

RoboTux commented:
> 
> Just spelling out what the regex means, because I've tripped myself up at least twice trying to interpret it. `^` means "start of line". `[^#]` means "any character except `#`. Combined they mean "match a non-empty line that does not start with `#`". `CHECK-NOT` instructs FileCheck to fail if it sees the specified between the previous thing it matched and the next thing it matches, so the combined effect is to fail the test if, between the previous and next patterns, a non-empty line _without_ a leading `#` appears. In other words, it will NOT fail if it sees e.g. `# error:` in the output, because of the leading `#` (and based on the checks elsewhere in the file, I assume that is indeed the format that errors will be printed in). Could somebody confirm that I'm not making a mistake?

That would be my reading as well.,Therefore I think "CHECK-NOT: # error: " is a better test but I think we should also split Inputs/shtest-env/* into a positive and negative tests and split shtest-env.py in two accordingly. This would allow to use RUN: %{lit} instead of RUN: not %{lit} thus guaranteeing that lit does not return an error.

Harini0924 commented:
@jh7370 @RoboTux Thank you for the feedback earlier. To clarify, I'm implementing the tests by splitting them into two separate test files—one for positive tests and one for negative tests.

- **Positive Tests:**  In the positive test file, I’m removing `not` from the RUN line (e.g., `# RUN: %{lit} -a -v %{inputs}/shtest-env-positive | FileCheck -match-full-lines %s`) and using `# CHECK-NOT: # error: `to ensure no errors occur.

- **Negative Tests:**  In the negative test file, I will keep `not` in the RUN line (e.g., `# RUN: not %{lit} -a -v %{inputs}/shtest-env-negative | FileCheck -match-full-lines %s`) and will use `# CHECK: # error: command failed with exit status: {{.*}}`, since these tests are expected to fail and will have an error line.

Does this approach align with what you were suggesting? I want to make sure I'm on the right track.

Harini0924 committed:
[llvm-lit] Split shtest-env tests into separate positive and negative test files

Created `shtest-env-positive.py` for positive tests, removing `not` from the RUN
line and using `CHECK-NOT: # error:` to ensure no errors occur. Created `shtest-env-negative.py`
for negative tests, keeping `not` in the RUN line and using `CHECK: # error: command failed with
exit status: {{.*}}` to verify expected failures. Split `tests/Inputs/shtest-env` into
`tests/Inputs/shtest-env-positive` and `tests/Inputs/shtest-env-negative`, organizing
the respective passing and failing tests into their appropriate directories.

Harini0924 committed:
[llvm-lit] Remove shtest-env files after restructuring

As part of the recent restructuring, the `shtest-env` directory
has been split into separate `shtest-env-positive` and
`shtest-env-negative` directories to better organize the
test cases. This commit removes the now obsolete files
from the original `shtest-env` directory that have been
migrated to the new structure.

jh7370 commented:
I'd like @RoboTux's feedback on the latest version before I do another round of reviewing this myself.

RoboTux commented:
> @jh7370 @RoboTux Thank you for the feedback earlier. To clarify, I'm implementing the tests by splitting them into two separate test files—one for positive tests and one for negative tests.
> 
>     * **Positive Tests:**  In the positive test file, I’m removing `not` from the RUN line (e.g., `# RUN: %{lit} -a -v %{inputs}/shtest-env-positive | FileCheck -match-full-lines %s`) and using `# CHECK-NOT: # error: `to ensure no errors occur.
> 
>     * **Negative Tests:**  In the negative test file, I will keep `not` in the RUN line (e.g., `# RUN: not %{lit} -a -v %{inputs}/shtest-env-negative | FileCheck -match-full-lines %s`) and will use `# CHECK: # error: command failed with exit status: {{.*}}`, since these tests are expected to fail and will have an error line.
> 
> 
> Additionally, I will be splitting `tests/Inputs/shtest-env` into `tests/Inputs/shtest-env-positive` and `tests/Inputs/shtest-env-negative`, with each containing their respective passing and failing tests.
> 
> Does this approach align with what you were suggesting? I want to make sure I'm on the right track.

Yes, that's exactly it. I'm not sure `CHECK-NOT: #error` is necessary for the positive case because I would expect lit to fail if any error is thrown but it cannot hurt.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env-args-none.txt:
```diff
@@ -0,0 +1,35 @@
+# Check default environment.
```
Nit: for newer tests, I tend to recommend using `##` for comments, to help them stand out from lit/FIleCheck lines. Since you're splitting the old test, this effectively becomes a new test, so it's a good chance to do that bit of tidying up.

Harini0924 replied:
Fixed.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/print_environment.py:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python
```
This isn't specific to the "positive" case, I assume? Could it be moved up out of the shtest-env-positive directory and used by both shtest-env-negative and shtest-env-positive?

ilovepi replied:
Is the file even needed now? AFAICT it was in place to work around `env` not working w/o a subcommand. Perhaps we should just remove it?

Harini0924 replied:
I agree that the `print_environment.py` file is no longer needed. I've integrated its functionality directly into the `env` command's logic. Now, when the `env` command is run without any arguments, it generates and returns a list of all environment variables in the `KEY=VALUE` format. This was the original purpose of `print_environment.py`, but since the `env` command itself now handles this, the script has become redundant and can be safely removed. What do you think?

jh7370 reviewed (commented):
Couple more suggestions, but this basically looks good.

I'm not convinced the `# error` check not lines are needed, but I'm also not opposed to them.

Harini0924 committed:
[llvm-lit] Removed `print_environment.py` file and Refactored some tests

This patch removes the print_environment.py file, which was previously
used to print environment variables in the format KEY=VALUE. The tests
have been refactored to directly use the env command, with the output
being validated using FileCheck. Comments were updated to use ## for
better readability, and the test cases were rewritten to ensure
consistency.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env-args-none.txt:
```diff
@@ -1,32 +1,32 @@
-# Check default environment.
+## Check default environment.
```
Thanks for improving the comments in this test, but the other test files modified/added in this PR need the same treatment.

While you're at it, could you add a comment at the very top of the file, separated with a blank line from first test case's comment, that explains the general purpose of the tests in this file, please. (Same goes for the other ones too).

Harini0924 replied:
I have added comments to the files I have modified. Is there anything else I should change?

Harini0924 committed:
[llvm-lit] Improved comments

Added a comment at the top of every file that I modified explaining the
general purpose of the tests in the file.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env-calls-env.txt:
```diff
@@ -1,25 +1,30 @@
-# Check that internal env can call internal env.
+## Tests env command's handling of nested calls, setting, unsetting, and variable retention.
```
Based on the file name and what the test cases actually do, I think this comment should say "Tests the behaviour of chaining env commands together." Or something to that effect anyway.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env.txt:
```diff
@@ -1,8 +1,10 @@
-# Check for simple one environment variable setting
+## Tests env command for setting single and multiple environment variables.
+
+## Check for simple one environment variable setting
```
Nit: here and below, missing trailing "."

Harini0924 replied:
I updated the comments in my previous commit.

jh7370 reviewed (commented):
Some nits, but nearly there.

Harini0924 committed:
Updated the comments in files modified

ilovepi commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env-args-none.txt:
```diff
@@ -1,4 +1,4 @@
-## Tests the env command's ability to handle setting, unsetting, and mixing environment variables.
+## Tests the env command in various scenarios: without arguments, setting, unsetting, and mixing envrionment variables. 
```
nit: trailing space

Harini0924 replied:
Done.

Harini0924 committed:
Fixed trailing space.

Harini0924 commented:
@jh7370 Are there any other changes I need to make?

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env-args-none.txt:
```diff
@@ -1,3 +1,5 @@
+## Tests the env command's ability to handle setting, unsetting, and mixing environment variables.
```
Please don't resolve comment threads I've started, even if you think you've addressed my comment. My comment was about the file name, which to me looks incorrect. What does "args-none" mean in the file name? `-u FOO` or whatever is not "no arguments".

jh7370 replied:
I think probably `env-no-subcommand.txt` would be the name that best-conveys its meaning. It looks like the common factor in these tests is that no additional process exists that is run with the env-specified variables.

Harini0924 replied:
I agree `env-no-subcommand.txt` is a better file name. I have renamed `env-args-none.txt` to `env-no-subcommand.txt` in my recent commit.

Harini0924 committed:
Renamed env-args-none.txt to env-no-subcommand.txt and alphabetized test order

jh7370 commented on llvm/utils/lit/tests/shtest-env-positive.py:
```diff
@@ -33,25 +33,24 @@
 # CHECK: PASS: shtest-env :: env-u.txt ({{[^)]*}})
 # CHECK: env -u FOO | {{.*}}
 # CHECK: # executed command: env -u FOO
 # CHECK-NOT: # error:
```
Why did the `# CHECK: --` line get removed?

Harini0924 replied:
I accidentally removed it in my last commit, but I added it back.

jh7370 commented on llvm/utils/lit/tests/Inputs/shtest-env-positive/env-no-subcommand.txt:
```diff
@@ -6,32 +6,32 @@
 # NO-ARGS: FOO=1
 # NO-ARGS: QUX=3
-
+   
```
Nit, you've added trailing whitespace again here and on many other lines in this file. Please fix!

Harini0924 replied:
Fixed.

Harini0924 committed:
Fixed trailing whitespaces and `# CHECK: --` which I accidentally removed in my last commit.

jh7370 reviewed (approved):
LGTM, assuming the pre-merge checks pass.

Harini0924 committed:
Merge branch 'main' into llvm_env
